### ISSUE TITLE
Bind NPC Functions

### ISSUE THREAD

**Issue by DrVrej** [Facepunch/garrysmod-requests#2502]
### Others / Hooks
- [ ] `NPC:EyePosition()` [Link](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_basenpc.h#L1405): This is something that people have wanted to be able to override for NPCs for a long time. Many NPCs in Half-life 2 override this as it controls the origin of the sight. Essentially when the eye position is accessed, whether through the engine or Lua (ENT:EyePos), it would allow NPCs that need to change their eye position to return a different value. If no value is returned or the call isn't accessed by the NPC, it would default back to the engine.
- [ ] `NPC:OnDoorFullyOpen(CBasePropDoor door)` [Link](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_basenpc.h#L1366): Called when the NPC finishes opening a door.
- [ ] `NPC:OnDoorBlocked(CBasePropDoor door)` [Link](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_basenpc.h#L1367): Called when the door that the NPC is opening gets blocked and doesn't finish opening.

### Getters / Setters
The following are variables from the engine that currently can be accessed through usage such as `GetInternalVariable`, however these are ran way too often in many NPC bases as they are essential, especially in schedule handling. This is an issue as using that lookup function is performance heavy. From tests done by couple of people and I, it's about 9-10x faster to access internal variables through a binded function. I have also provided links to for each one to the corresponding place in the engine.

- [ ] `NPC:GetMoveWaitTime()` [Link](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_basenpc.h#L1360): Returns the value of `m_flMoveWaitFinished`, a float value.
- [ ] `NPC:SetMoveWaitTime(float time)` [Link](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_basenpc.h#L1360): Sets the value of `m_flMoveWaitFinished`.
- [ ] `NPC:GetLastPosition()` [Link](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_basenpc.h#L1854): Returns the value of `m_vecLastPosition`, a vector value. Currently we have the setter: [NPC:SetLastPosition](https://wiki.facepunch.com/gmod/NPC:SetLastPosition).
- [ ] `NPC:GetLastDamageTime()`  [Link](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_basenpc.h#L1547): Returns the value of `m_flLastDamageTime`, a float value.
- [ ] `NPC:GetOpeningDoor()` [Link](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_basenpc.h#L1368): Returns the value of `m_hOpeningDoor`, the door entity or NULL.

**Comment by DrVrej** [comment]
> along with NPC.EyePosition, hooking NPC.BodyTarget and NPC.HeadTarget would be better as well. HL2 NPCs use these to get target coordinates for enemy, and they also use this to modify their target spots without manipulating eye position. 

Yup, a lot of NPC bases use that as well including VJ Base, would be helpful to have them hookable alongside eye position. In fact some HL2 NPCs modify it to change their target position

**Comment by Roachdaripper** [comment]
bump

**Comment by Cpt-Hazama** [comment]
bump

**Comment by robotboy655** [comment]
Added `SetMoveWaitTime` as `NPC:SetMoveDelay`, it automatically adds curtime. Same for GetMoveDelay. Let me know if this is a bad idea, but every usage of it in base game deals with curtime, so I thought it'd be better to deal with it behind the scenes.

I am hesitant on adding hooks for those things, as it will most definitely slow down every Lua NPC, even when they don't use said hooks because they would be called multiple times a tick per each NPC. You should be already able to overwrite the "eye" pos via https://wiki.facepunch.com/gmod/NPC:SetViewOffset, right?

**Comment by DrVrej** [comment]
@robotboy655 Thank you for adding those, having the CurTime within the function is better so we don't have to call it through Lua 👍

### **Eye Position & Direction:**
There is 3 main issues with using [SetViewOffset](https://wiki.facepunch.com/gmod/NPC:SetViewOffset).
**ISSUE 1:**
If the NPC's model has the `eyes` attachment, it automatically overrides both the eye position and the direction. It does this constantly by calling `CAI_BaseActor::UpdateLatchedValues` in `ai_baseactor` as seen here: https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/server/ai_baseactor.cpp#L700. This is great if we want to use the `eyes` attachment, however it means that we can't edit it at all whether by using `SetViewOffset` or changing the internal variable `m_vDefaultEyeOffset`

**ISSUE 2:**
If the NPC's model does NOT have `eyes` attachment, it reverts to using the current animation's eye position. This means if we want to override it, we have to constantly call `SetViewOffset` and also set the internal variable `m_vDefaultEyeOffset` on the NPC's `Think` function. However this is still an issue because it randomly offsets as seen in example 2 below.

**ISSUE 3:**
We have no way of changing the direction of the eye position, it's either determined by the `eyes` attachment or set to the NPC's forward direction. So in cases where the direction should be following the custom view offset, it doesn't as seen in example 2 below.

### **EXAMPLES**
**EXAMPLE 1:**
NPC has an `eyes` attachment but it's facing the wrong direction. We have no way of fixing without decompiling the model and editing the attachment's rotation:
![Image](https://github.com/user-attachments/assets/4a96d413-628b-4caf-a272-b251297efaeb)

**EXAMPLE 2:**
NPC has no `eyes` attachment, so we are running the following script constantly in the NPC's `Think` function:
```
local newEyePos = self:WorldToLocal(self:GetAttachment(self:LookupAttachment("mouth")).Pos)
self:SetViewOffset(newEyePos)
self:SetSaveValue("m_vDefaultEyeOffset", newEyePos)
```
We can see both issue 2 and 3 with this. The position is randomly offset to the side even though it's supposed to be at the model's mouth attachment, and since we don't have a way to set the direction, there is no way to tell it to follow the mouth attachment's forward direction.
![Image](https://github.com/user-attachments/assets/7e94527e-fdcf-4b38-9948-0443d223f13d)

### **Conclusion**
The performance loss is an important concern, so maybe instead of hook we could have a variable that can override the hard coded `eyes` attachment name to another attachment name? This will at least solve some of the issues as I listed above. It could be named as `NPC:SetEyePosAttachment` or something along those lines.
